devilspie2.1
.\"
.\" Copyright © 2011-2017 Andreas Rönnquist.
.\" Copyright © 2021-2024 Darren Salt.
.\" This file is distributed under the same license
.\" as the devilspie2 package, see COPYING file.
.\"
.TH devilspie2 1 "" ""
.SH NAME
Devil's Pie 2 \- perform scripted actions on windows when certain events occur
.SH Description
.P
.B Devil's Pie 2
is a program that detects windows as they are created, gain or lose focus or
have their titles changed, and performs scripted actions on them. The
scripts are written in \fILua\fR, allowing a great deal of customisation.

.SH Options
.TP
\fB\-v\fR, \fB\-\-version
Displays the version of Devil's Pie 2.
.TP
\fB\-f \fIfolder\fR, \fB\-\-folder \fIfolder
Sets the folder where the Lua scripts are loaded from.
.TP
\fB\-d\fR, \fB\-\-debug
Shows debug information from the Lua scripts. If \fIdebug_print\fR is used
in the Lua scripts, it will output to stdout.
.TP
\fB\-D\fR, \fB\-\-debug\-fifo
Copy stdout & send debug info to a FIFO. If \fIdebug_print\fR is used in
the Lua scripts, it will output to the FIFO. This is intended for running
dæmonised while allowing access to debug text etc. at need.

Nothing is output to the FIFO if nothing is reading it.
You can read it using a program such as \fBcat\fR.

\fB\-\-debug\fR and \fB\-\-debug\-fifo\fR may be used together.
.TP
\fB\-P\fR, \fB\-\-print\-fifo
Print the FIFO's file name then exit. This may be used as follows:
.EX
cat "$(devilspie2 \-\-print\-fifo)"
.EE
.TP
\fB\-e\fR, \fB\-\-emulate
Emulation mode. This prevents windows from being affected by the scripts,
but window positions etc. can still be read.
.TP
\fB\-w\fR, \fB\-\-wnck\-version
Show the version of libwnck in use. (Only available on GTK3 or later.)
.TP
\fB\-l\fR, \fB\-\-lua\-version
Show the version of Lua in use.
.SH Script Commands
.P
See
.I /usr/share/doc/devilspie2/README.md
for a detailed description of the commands recognised in Lua scripts.

.SH Files
.TP
.B $XDG_RUNTIME_DIR/devilspie2\-$DISPLAY
.TP
.B $TMPDIR/devilspie2\-$DISPLAY
.TP
.B /tmp/devilspie2\-$DISPLAY
These are the possible locations of the FIFO which is created if
\fB\-\-debug\-fifo\fR is used. The first one for which the referenced
environment variables are set is used.

.SH Author
.P
.B Devil's Pie 2
is maintained by 
.MT devspam@moreofthesa.me.uk
Darren Salt
.ME . It was written by
.MT andreas@ronnquist.net
Andreas Rönnquist
.ME , and is based on Devil's Pie by Ross
Burton.


ChangeLog

ChangeLog:
----------

0.46
	* Converted README files to Markdown.
	* Added a time-out for scripts.
	  (Release build only; inactive if DEBUG is defined at build time.)
	* Improved parameter checking (and better error messages).

0.45
	* Fixes related to Lua version handling
	* Misc build fixes
	* Add set_on_bottom(): move to bottom of stack
	* Change set_on_top() so that it merely raises the window
	* Add get_class_group_name().
          This complements get_class_instance_name re. WM_CLASS.
	* More aliases for function names containing 'maximize' or
	  'minimize': accept -ise spellings
	* Visibly deprecate function names containing 'horisontally'
	* Add set_adjust_for_decoration(), useful if you have problems with
	  positioning & resizing of decorated windows [note: experimental]
	* Add get_window_frame_extents()
	* Allow workspaces to be referenced by name
	* Extra event type for running rules on window name/title changes
	* Ignore dot files in the scripts directory.
	* More UTF-8 support: use_utf8() (sets and reads whether UTF-8 is
	  default, initially false), window_property_is_utf8() and
	  window_property_full().

0.44
	* Default to Lua 5.3, and permit use of other versions
	* Add support for setting and deleting window properties
	* Add get_process_name(), which returns the name of the process
	  owning the window (if possible)
	* Add get_monitor_index() and get_monitor_geometry(); improve
	  monitor index/geometry discovery
	* Add get_window_is_decorated() (which uses the Motif window hints)
	* Alias center() as centre()
	* Extend centre() to allow centring on a given monitor and to allow
	  horizontal or vertical only
	* Extend set_window_position() to allow positioning on a given
	  monitor and control which corner is measured from.
	  (Behaviour with two parameters is unchanged.)
	* Quote the option hyphens in the man page (preventing replacement)
	* Various const & compiler warning (gcc -Wextra) fix-ups
	* Support for UTF-8 strings in window properties
	* Allow replacements for pkg-config at compile time

0.43
	* Update contact adresses
	* Move homepage to nongnu.org

0.42
	* Support window focus / blur events
	* Add example for Awesome window manager
	* Updated French, Italian and Swedish translations

0.41
	* Set a correct version number

0.40
	* Add callback for geometry change
	* Add error string for one or two indata expected
	* Make devilspie2_get_viewport_start return both x
	  and y coordinates
	* Add alternative functionality to set_viewport Now you can set the
	  coordinates to the window if you use two integers as indata
	* New Italian translation, updated Dutch, Japanese and Swedish
	  translations
	* Build flags hardening fixes
	* Allow reading of XA_WINDOW properties

0.39
	* Fix proper flushing of debugging output
	* Some source formatting fixes
	* add get_window_is_pinned function
	* Make buildable with a custom CC like clang

0.38
	* Add documentation of xy and xywh functions
	* Source cleanups
	* Move setting of current window variable

0.37
	* Replace google group with savannah mailinglist
	* Add xy and xywh functions

0.36
	* _Really_ fix the man page commenting style
	* Remove version and date from man-page

0.35
	* Fix man page commenting

0.34
	* Add GPL headers in more places where needed
	* Fix cppcheck reassignment warning
	* Install man-page with Makefile too

0.33
	* Add boolean argument to set_window_{above,below} (Thanks to rbtwrrr)
	* Fix spelling of horizontally in all places
	* Various source formatting fixes

0.32
	* Add functions get_fullscreen (get_window_fullscreen)
	* Small documentation fixes
	* Add set_window_strut function
	* gdk_display_get_n_screens is deprecated on GLIB >= 3.10, there we default
	  to only one screen.
	* Updated Russian translation

0.31
	* Add close_window function
	* Remove unused C function my_wnck_get_cardinal
	* Fix string problems reported by pscan
	* Updated Russian translation
	* Added Japanese translation

0.30
	* Update French translation
	* Add function set_window_opacity to do the same as set_opacity
	* Add maximize_horizontally with same functionality as
	  maximize_horisontally
	* Add get_window_is_maximized_horizontally with same functionality as
	  get_window_is_maximized_horisontally
	* Add focus and focus_window functions
	* Updated Brasilian-Portugese translation

0.29
	* Add finnish translation
	* Updated dutch translation
	* Add examples to get_window_geometry and get_window_client_geometry
	* Add get_screen_geometry function (Thanks to Ronan Barzic)

0.28
	* Make get workspace error string consistent with other past-tense strings
	* wnck_class_group_get_res_class is deprecated, use wnck_class_group_get_id
	  instead on wnck versions it isn't available
	* Make 'relative window positioning' possible, allowing negative values to
	  set_window_geometry
	* Remove sciteproj project file
	* Add set_window_type function
	* Add set_window_opacity function
	* Added handling of window closing events
	* Add file monitor to automatically load added scripts

0.27
	* Update Makefile to fix a Lua problem with pkg-config
	* Update the email contact address in the README
	* Add get_workspace_count function - Thanks Hans Nieser!
	* Use luaL_newstate instead of lua_open which is deprecated

0.26
	* Add "center" function
	* Register set_window_position2 under the correct name - Fixes
	  bug where set_window_position2 couldn't be found
	* Add a text file describing development for people who
	  want to join
	* Add a TODO file
	* Updated translations

0.25
	* Add set_viewport function
	* Move error strings to a source file of their own
	* Updated translations - Thanks guys!
	* Make it possible to build using -j without screwing up
	  translation build

0.24
	* Fix bug where folder settings got overridden
	  - Thanks Andreas Müller!
	* Add set_window_fullscreen function
	* Add some required checks for running in emulation mode

0.23
	* Added Russian translation - Thanks to Roman V. Nikolaev

src/devilspie2.c
/**
 *	This file is part of devilspie2
 *	Copyright (C) 2005 Ross Burton, 2011-2017 Andreas Rönnquist
 *
 *	devilspie2 is free software: you can redistribute it and/or
 *	modify it under the terms of the GNU General Public License as published
 *	by the Free Software Foundation, either version 3 of the License, or
 *	(at your option) any later version.
 *
 *	devilspie2 is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *	GNU General Public License for more details.
 *
 *	You should have received a copy of the GNU General Public License
 *	along with devilspie2.
 *	If not, see <http://www.gnu.org/licenses/>.
 */

#include <string.h>

#include <stdlib.h>
#include <glib.h>
#include <glib/gstdio.h>
#include <gdk/gdk.h>
#include <gdk/gdkx.h>
#include <glib/gi18n.h>

#define WNCK_I_KNOW_THIS_IS_UNSTABLE
#include <libwnck/libwnck.h>

#include <lua.h>
#include <lualib.h>
#include <lauxlib.h>

#include <locale.h>

#include "script.h"
#include "script_functions.h"
#include "logger.h"

#include "error_strings.h"

#include "config.h"


#if (GTK_MAJOR_VERSION >= 3)
#define HAVE_GTK3
#endif

/**
 *
 */
GMainLoop *loop = NULL;

static gboolean debug = FALSE;
static gboolean logtofifo = FALSE;
static gboolean emulate = FALSE;

static gboolean show_fifo = FALSE;

static gboolean show_version = FALSE;

// libwnck Version Information is only availible if you have
// libwnck 3.0 or later
static gboolean show_wnck_version = FALSE;

static gboolean show_lua_version = FALSE;

static gchar *script_folder = NULL;
static gchar *temp_folder = NULL;

GFileMonitor *mon = NULL;

gchar *config_filename = NULL;

WnckHandle *my_wnck_handle = NULL;

/**
 *
 */
static void load_list_of_scripts(WnckScreen *screen G_GNUC_UNUSED, WnckWindow *window,
                                 GSList *file_list)
{
	GSList *temp_file_list = file_list;
	// set the window to work on
	set_current_window(window);

	// for every file in the folder - load the script
	while(temp_file_list) {
		gchar *filename = (gchar*)temp_file_list->data;

		// is it a Lua file?
		if (g_str_has_suffix((gchar*)filename, ".lua")) {

			// init the script, run it
			if (!run_script(global_lua_state, filename))
				/**/;

		}
		temp_file_list=temp_file_list->next;
	}
	return;

}


static void window_name_changed_cb(WnckWindow *window)
{
	WnckScreen * screen = wnck_window_get_screen(window);
	if(screen == NULL) return;

	// Handle duplicate name-change events
	// Simple method: just track the most recent event regardless of window
	static WnckWindow *previous = NULL;
	static char *prevname = NULL;

	const char *newname = wnck_window_get_name(window);
	if (window == previous && prevname && !strcmp (prevname, newname))
		return;
	// Store the info for the next event
	free(prevname);
	prevname = strdup(newname);
	previous = window;

	load_list_of_scripts(screen, window, event_lists[W_NAME_CHANGED]);
}

/**
 *
 */
static void window_opened_cb(WnckScreen *screen, WnckWindow *window)
{
	load_list_of_scripts(screen, window, event_lists[W_OPEN]);
	/*
	Attach a listener to each window for window-specific changes
	Safe to do this way as long as the 'user data' parameter is NULL
	*/
	g_signal_connect(window, "name-changed", (GCallback)window_name_changed_cb, NULL);
}


/**
 *
 */
static void window_closed_cb(WnckScreen *screen, WnckWindow *window)
{
	load_list_of_scripts(screen, window, event_lists[W_CLOSE]);
}


/**
 *
 */
static void window_changed_cb(WnckScreen *screen, WnckWindow *window)
{
	WnckWindow *cur;

	load_list_of_scripts(screen, window, event_lists[W_BLUR]);
	cur = wnck_screen_get_active_window(screen);
	load_list_of_scripts(screen, cur, event_lists[W_FOCUS]);
}


/**
 *
 */
void init_screens()
{
	int i;
	int num_screens;

#ifndef GDK_VERSION_3_10
	num_screens = gdk_display_get_n_screens(gdk_display_get_default());
#else
	num_screens = 1;
#endif

	for (i=0; i<num_screens; i++) {
		WnckScreen *screen = wnck_handle_get_screen(my_wnck_handle, i);

		g_signal_connect(screen, "window-opened",
		                 (GCallback)window_opened_cb, NULL);
		g_signal_connect(screen, "window-closed",
		                 (GCallback)window_closed_cb, NULL);
		g_signal_connect(screen, "active-window-changed",
		                 (GCallback)window_changed_cb, NULL);
	}
}


/**
 * atexit handler - kill the script
 */
void devilspie_exit()
{
	clear_file_lists();
	g_free(temp_folder);
	if (mon)
		g_object_unref(mon);
	g_free(config_filename);
}


/**
 * handle signals that are sent to the application
 */
static void signal_handler(int sig)
{
	printf("\n%s %d (%s)\n", _("Received signal:"), sig, strsignal(sig));

	done_script_error_messages();

	if (sig == SIGINT) {
		exit(EXIT_FAILURE);
	}
}


/**
 *
 */
static void print_list(GSList *list)
{
	GSList *temp_list;
	if (list != NULL) {
		temp_list = list;

		while(temp_list) {
			gchar *file_name = temp_list->data;

			if (file_name) {
				if (g_str_has_suffix((gchar*)file_name, ".lua")) {
					logger_printf("%s\n", (gchar*)file_name);
				}
			}
			temp_list = temp_list->next;
		}
	}
}


/**
 *
 */
static void print_script_lists()
{
	gboolean have_any_files = FALSE;
	win_event_type i;

	logger_print("------------\n");

	for (i = 0; i < W_NUM_EVENTS; i++) {
		if (event_lists[i])
			have_any_files = TRUE;
		// If we are running debug mode - print the list of files:
		logger_printf(_("List of Lua files handling \"%s\" events in folder:\n"),
		              event_names[i]);
		if (event_lists[i]) {
			print_list(event_lists[i]);
		}
	}

	if (!have_any_files) {
		logger_err_printf("%s\n\n", _("No script files found in the script folder - exiting."));
		exit(EXIT_SUCCESS);
	}
}

/**
Reload the config and re-initialise the Lua VM
 */
void refresh_config_and_script()
{
	clear_file_lists();
	set_current_window(NULL);
	
	if (load_config(config_filename) != 0) {
		logger_print_always("Configuration file cannot be re-loaded. Processing will not continue until the error is corrected.\n");
		return;
	}
	
	global_lua_state = reinit_script(global_lua_state, script_folder);

	logger_print("Files in folder updated!\n - new lists:\n\n");

	print_script_lists();

	logger_print("-----------\n");
}

/**
 *
 */
#ifdef FILENAME_MAX
# define PATHLEN (FILENAME_MAX)
#elif defined(PATH_MAX)
# define PATHLEN (PATH_MAX)
#else
# define PATHLEN 4096
#endif
static struct {
	guint ref;
	gboolean refresh; // true if devilspie2.lua is changed or if a file is created or deleted
	char *short_filename; // NULL to force full refresh
	char full_path[PATHLEN];
} folder_changed_data;

static void folder_changed_callback_idle(gpointer user_data)
{
printf("callback: refresh=%s name=%s short=%s\n", folder_changed_data.refresh?"TRUE ":"FALSE", folder_changed_data.full_path, folder_changed_data.short_filename);
	if (folder_changed_data.refresh || !folder_changed_data.short_filename)
	{
		refresh_config_and_script();
	}
	else if( g_str_has_suffix((gchar*)folder_changed_data.short_filename, ".lua") && is_in_any_list(folder_changed_data.full_path) == FALSE)
	{	// May be a module file 
puts("lua, in list");
		gchar * module_name = g_utf8_substring(folder_changed_data.short_filename, 0, strlen(folder_changed_data.short_filename) - 4);
		if(is_module_loaded(global_lua_state, module_name) == TRUE)
		{
			global_lua_state = reinit_script(global_lua_state, script_folder);
puts("reinit");
		}
		g_free(module_name);
	}
	if (!user_data)
		folder_changed_data.ref = 0;
}

void folder_changed_callback(GFileMonitor *mon G_GNUC_UNUSED,
                             GFile *first_file,
                             GFile *second_file G_GNUC_UNUSED,
                             GFileMonitorEvent event,
                             gpointer user_data)
{
	if (!folder_changed_data.ref) {
		// reset if the callback isn't registered
		folder_changed_data.refresh = FALSE;
		folder_changed_data.short_filename = NULL;
		folder_changed_data.full_path[0] = 0;
	}

	// If a file is created or deleted, we need to check the file lists again
	if ((event == G_FILE_MONITOR_EVENT_CREATED) ||
	    (event == G_FILE_MONITOR_EVENT_DELETED))
	{
		// created or deleted
		folder_changed_data.refresh = TRUE;
		folder_changed_data.short_filename = NULL;

		goto add_callback;
	}

	// Also monitor if our devilspie2.lua file is changed - since it handles
	// which files are window close or window open scripts.
	if (event == G_FILE_MONITOR_EVENT_CHANGED && first_file && !folder_changed_data.refresh)
	{
		// assumption: short_filename is trailing substr of full_path
		gchar *short_filename = g_file_get_basename(first_file);
		gchar *full_path = g_file_get_path(first_file);

		if (g_strcmp0(short_filename, "devilspie2.lua") == 0)
			folder_changed_data.refresh = TRUE;
		// paths differ? process the previous one now
		else if (folder_changed_data.short_filename && g_strcmp0(full_path, folder_changed_data.full_path))
		{
			folder_changed_callback_idle(&folder_changed_data); // non-NULL is all that's needed
		}

		folder_changed_data.short_filename = folder_changed_data.full_path + strlen(full_path) - strlen(short_filename);
		strncpy(folder_changed_data.full_path, full_path, PATHLEN - 1);
		folder_changed_data.full_path[PATHLEN - 1] = 0;

		g_free(short_filename);
		g_free(full_path);

		goto add_callback;
	}
	// not an event in which we're interested; return without adding the callback
	return;

add_callback: // add the callback
	if (!folder_changed_data.ref)
		folder_changed_data.ref = g_idle_add_once(folder_changed_callback_idle, NULL);
}

/**
 * Program main entry
 */
int main(int argc, char *argv[])
{
	static const GOptionEntry options[]= {
		{ "debug",        'd', 0, G_OPTION_ARG_NONE,   &debug,
		  N_("Print debug info to stdout"), NULL
		},
		{ "debug-fifo",   'D', 0, G_OPTION_ARG_NONE,   &logtofifo,
		  N_("Print debug info & copy stdout to a FIFO"), NULL
		},
		{ "print-fifo",   'P', 0, G_OPTION_ARG_NONE,   &show_fifo,
		  N_("Print the debug FIFO's file name then quit"), NULL
		},
		{ "emulate",      'e', 0, G_OPTION_ARG_NONE,   &emulate,
		  N_("Don't apply any rules, only emulate execution"), NULL
		},
		{ "folder",       'f', 0, G_OPTION_ARG_STRING, &script_folder,
		  N_("Search for scripts in this folder"), N_("FOLDER")
		},
		{ "version",      'v', 0, G_OPTION_ARG_NONE,   &show_version,
		  N_("Show Devilspie2 version and quit"), NULL
		},
		// libwnck Version Information is only availible if you have
		// libwnck 3.0 or later
		{ "wnck-version", 'w', 0, G_OPTION_ARG_NONE,   &show_wnck_version,
		  N_("Show libwnck version and quit"), NULL
		},
		{ "lua-version",  'l', 0, G_OPTION_ARG_NONE,   &show_lua_version,
		  N_("Show Lua version and quit"), NULL
		},
		{ NULL }
	};

	GError *error = NULL;
	GOptionContext *context;

	// Init gettext stuff
	setlocale(LC_ALL, "");

	bindtextdomain(PACKAGE, LOCALEDIR);
	bind_textdomain_codeset(PACKAGE, "");
	textdomain(PACKAGE);

	gchar *devilspie2_description = g_strdup_printf(_("apply rules on windows"));

	gchar *full_desc_string = g_strdup_printf("- %s", devilspie2_description);

	context = g_option_context_new(full_desc_string);
	g_option_context_add_main_entries(context, options, NULL);
	if (!g_option_context_parse(context, &argc, &argv, &error)) {
		g_print(_("option parsing failed: %s"), error->message);
		printf("\n");
		exit(EXIT_FAILURE);
	}

	if (show_fifo) {
		puts(logger_get_fifo_name());
		exit(EXIT_SUCCESS);
	}

	gboolean shown = FALSE;
	if (show_version) {
		printf("Devilspie2 v%s\n", DEVILSPIE2_VERSION);
		shown = TRUE;
	}
	// libwnck Version Information is only availible if you have
	// libwnck 3.0 or later
	if (show_wnck_version) {
#ifdef _DEBUG
		printf("GTK v%d.%d.%d\n",
		       GTK_MAJOR_VERSION,
		       GTK_MINOR_VERSION,
		       GTK_MICRO_VERSION);
#endif
#ifdef HAVE_GTK3
		printf("libwnck v%d.%d.%d\n",
		       WNCK_MAJOR_VERSION,
		       WNCK_MINOR_VERSION,
		       WNCK_MICRO_VERSION);
#else
		printf("libwnck v2.x\n");
#endif
		shown = TRUE;
	}
	if (show_lua_version) {
		puts(LUA_VERSION);
		shown = TRUE;
	}
	if (shown)
		exit(0);

	gdk_init(&argc, &argv);

	g_free(full_desc_string);
	g_free(devilspie2_description);
	g_option_context_free(context);
	
	// if the folder is NULL, default to ~/.config/devilspie2/
	if (script_folder == NULL) {

		temp_folder = g_build_path(G_DIR_SEPARATOR_S,
		                           g_get_user_config_dir(),
		                           "devilspie2",
		                           NULL);

		// check if the folder does exist
		if (!g_file_test(temp_folder, G_FILE_TEST_IS_DIR)) {

			// - and if it doesn't, create it.
			if (g_mkdir(temp_folder, 0700) != 0) {
				printf("%s\n", _("Couldn't create the default folder for devilspie2 scripts."));
				exit(EXIT_FAILURE);
			}
		}

		script_folder = temp_folder;
	}


#if (GTK_MAJOR_VERSION >= 3)
	if (!GDK_IS_X11_DISPLAY(gdk_display_get_default())) {
		puts(_("An X11 display is required for devilspie2."));
		if (getenv("WAYLAND_DISPLAY"))
			puts(_("Wayland & XWayland are not supported.\nSee https://github.com/dsalt/devilspie2/issues/7"));
		puts("");
		return EXIT_FAILURE;
	}
#endif

	if (init_script_error_messages()!=0) {
		printf("%s\n", _("Couldn't init script error messages!"));
		exit(EXIT_FAILURE);
	}

	// set the current window to NULL, we don't need to be able to modify
	// the windows when reading the config
	set_current_window(NULL);

	config_filename =
	    g_build_filename(script_folder, "devilspie2.lua", NULL);

	if (load_config(config_filename) != 0) {

		devilspie_exit();
		return EXIT_FAILURE;
	}

	if (debug) {

		if (emulate) {
			printf("%s\n\n", _("Running devilspie2 in debug and emulate mode."));
		} else {
			printf("%s\n\n", _("Running devilspie2 in debug mode."));
		}

		printf(_("Using scripts from folder: %s"), script_folder);

		printf("\n");

		devilspie2_debug = TRUE;
	}

	// Should we only run an emulation (don't modify any windows)
	if (emulate) devilspie2_emulate = emulate;

	GFile *directory_file;
	directory_file = g_file_new_for_path(script_folder);
//	mon = g_file_monitor_directory(directory_file, G_FILE_MONITOR_WATCH_MOUNTS,
	mon = g_file_monitor_directory(directory_file, G_FILE_MONITOR_NONE,
	                               NULL, NULL);
	g_object_unref(directory_file);
	if (!mon) {
		printf("%s\n", _("Couldn't create directory monitor!"));
		return EXIT_FAILURE;
	}

	g_signal_connect(mon, "changed", G_CALLBACK(folder_changed_callback),
	                 (gpointer)(config_filename));

	global_lua_state = init_script(script_folder);
	if (logtofifo)
		logger_create(global_lua_state);
	print_script_lists();

	logger_print("------------\n");

	// remove stuff cleanly
	atexit(devilspie_exit);

	struct sigaction signal_action;

	sigemptyset(&signal_action.sa_mask);
	signal_action.sa_flags = 0;
	signal_action.sa_handler = signal_handler;

	if (sigaction(SIGINT, &signal_action, NULL) == -1) {
		exit(EXIT_FAILURE);
	}

	my_wnck_handle = wnck_handle_new(WNCK_CLIENT_TYPE_PAGER);
	init_screens();

	loop=g_main_loop_new(NULL, TRUE);
	g_main_loop_run(loop);

	return EXIT_SUCCESS;
}
